File: dms613412/Programmars/Level2/[2020 카카오 공채] 괄호변환.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

string u, v, answer = "";


string change(string u) {
	u = u.substr(1, u.size() - 2);
	string rel = "";
	for (int i = 0; i < u.size(); ++i) {
		if (u[i] == '(')
			rel += ")";
		else
			rel += "(";
	}
	cout << rel << endl;
	return rel;
}

bool balance(string u) {
	if (u[0] == ')')return false;
	int left = 1;
	for (int i = 1; i < u.size(); ++i) {
		if (u[i] == '(')++left;
		else if (--left < 0) return false;
	}
	if (left == 0) return true;
}

string solution(string p) {
	if (p == "")
		return p;

	int left = 0, right = 0, slash = 0;
	for (int i = 0; i < p.size(); ++i) {
		if (p[i] == '(')++left;
		else ++right;
		if (left == right) {
			slash = left + right - 1;
			break;
		}
	}

	u = p.substr(0, slash + 1);
	v = p.substr(slash + 1);
	if (balance(u)) {
		answer += u;
		solution(v);
	}
	else {
		string v_ = v, u_ = u;
		answer += "(";
		solution(v);
		v = v_; u = u_;
		answer += ")";
		answer += change(u);
	}

	return answer;
}

File: dms613412/Programmars/Level2/[2020 카카오 공채] 괄호변환.py
def change(str):
    rel = ""
    for i in range(1,len(str)-1):
        if(str[i]=='('): 
            rel += ')'
        else: rel += "("
    return rel

def balance(str):
    if(str[0]==')'): 
        return False
    left = 0
    for i,val in enumerate(str):
        if(val=="("): 
            left+=1
        else:
            left-=1            
            if(left<0): return False
    if(left == 0): return True

def findSlash(str):
    num =0
    for idx, value in enumerate(str):
        if value == ")":
            num -=1
        if value == "(":
            num +=1
        if num == 0:
            return idx
    
def solution(p):
    answer = ''
    if(len(p)==0):
        return p

    slash = findSlash(p)        
    u = p[:slash+1]
    v = p[slash+1:]
    if(balance(u)):
        return u + solution(v)
    else:
        v_=v
        u_=u
        answer += '(' + solution(v)
        v=v_
        u=u_
        answer += ')' + change(u)
        return answer 
    return answer
